From c397e5e13c53f194655b873ffe643f7ea91b83b3 Mon Sep 17 00:00:00 2001
From: Murali Karicheri <m-karicheri2@ti.com>
Date: Wed, 4 Mar 2020 23:33:34 +0000
Subject: [PATCH 20/25] net: ti: icssg-prueth: Add shutdown command

As part of prueth_stop(), stop the firmware packet processing
by issuing a shutdown command so that interface is taken down
gracefully.

Signed-off-by: Murali Karicheri <m-karicheri2@ti.com>
Acked-by: Roger Quadros <rogerq@ti.com>
---
 drivers/net/ti/icssg-prueth.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/net/ti/icssg-prueth.c b/drivers/net/ti/icssg-prueth.c
index 5d5b932f63..5abb221601 100644
--- a/drivers/net/ti/icssg-prueth.c
+++ b/drivers/net/ti/icssg-prueth.c
@@ -71,6 +71,12 @@ enum prueth_port {
 /* CTRLMMR_ICSSG_RGMII_CTRL register bits */
 #define ICSSG_CTRL_RGMII_ID_MODE		BIT(24)
 
+/* Shutdown command to stop processing at firmware.
+ * Command format : 0x8101ss00. ss - sequence number. Currently not used
+ * by driver.
+ */
+#define ICSSG_SHUTDOWN_CMD		0x81010000
+
 /* pstate speed/duplex command to set speed and duplex settings
  * in firmware.
  * Command format : 0x8102ssPN. ss - sequence number: currently not
@@ -539,6 +545,10 @@ static void prueth_stop(struct udevice *dev)
 
 	icssg_class_disable(priv->miig_rt, priv->slice);
 
+	/* Execute shutdown command at firmware */
+	if (icssg_execute_firmware_command(priv, ICSSG_SHUTDOWN_CMD))
+		dev_err(dev, "Error executing firmware shutdown cmd\n");
+
 	phy_shutdown(priv->phydev);
 
 	dma_disable(&priv->dma_tx);
-- 
2.17.1

